### DISCUSSION TITLE
add css class to table?

### DISCUSSION THREAD

**Discussion by madebyafox** [quarto-dev/quarto-cli#849] [Q&A]
Loving Quarto!! 

Q: I see I can specify css classes and styles to divs and imgs added through the Visual Editor... but I'm not seeing this options for tables. Am I missing this somewhere?

(use case) I have a number of markdown tables used to organize images in columns, and want to set column widths for different 'types' of tables using css. 

Thanks!

**Comment by jjallaire** [comment]
We don't currently have a way to set classes on tables however we do have a way to set explicit column widths (see https://quarto.org/docs/authoring/tables.html#column-widths)

**Reply by jmgirard** [reply]
It would be nice to be able to do something like:

```
| A | B | C |
|---|---|---|
| 1 | 2 | 3 |
: {.css-classes} 
```

**Reply by cscheid** [reply]
Note that you should still be able to wrap your table in a div with a class, and then use CSS selectors:

```

::: {.css-classes}

| A | B | C |
|---|---|---|
| 1 | 2 | 3 |

:::

```

**Reply by fuhrmanator** [reply]
With bootstrap/bootswatch (the themes that Quarto supports), you need to have the classes in the table itself (a div doesn't work). A good example is the one from the documentation at https://quarto.org/docs/authoring/tables.html#tbl-planets (but it was generated by Python!). It has a `<table class="table table-sm table-striped">`. 

I see that Quarto HTML tables already have the `table` class, so it would be cool as @jmgirard suggested to allow for adding the other bootstrap classes. Of course, it's not obvious how to get rows and cells to have their appropriate classes, but it's a start (striped tables would already be cool).

**Reply by cscheid** [reply]
This lua filter should work for @jmgirard's suggested syntax:

```lua
-- is the table a simple array?
-- see: https://web.archive.org/web/20140227143701/http://ericjmritz.name/2014/02/26/lua-is_array/
function tisarray(t)
  local i = 0
  for _ in pairs(t) do
      i = i + 1
      if t[i] == nil then return false end
  end
  return true
end

-- https://stackoverflow.com/a/7615129
local function mysplit (inputstr, sep)
  if sep == nil then
    sep = "%s"
  end
  local t={}
  for str in string.gmatch(inputstr, "([^"..sep.."]+)") do
    table.insert(t, str)
  end
  return t
end -- note that in this function, .foo.bar splits the same as foo.bar

-- this won't match { .classes } or {.class1 .class2}
-- it only matches {.class}, {.class1.class2}, etc. NO spaces!
function Table(tbl)
  local function find_text_content(node)
    if node.t == "Plain" then
      return node.content
    end
    if node.t == "Para" then
      return node.content
    end
    if tisarray(node) then
      return find_text_content(node[1])
    end
    if node.t == "BlockQuote" then
      return find_text_content(node.content[1])
    end
    if node.t == "CodeBlock" then
      return node.content
    end
    if node.t == "Header" then
      return node.content
    end
    if node.t == "HorizontalRule" then
      return nil
    end
    if node.t == "LineBreak" then
      return nil
    end
    if node.t == "Null" then
      return nil
    end
    if node.t == "Space" then
      return nil
    end
    if node.t == "Str" then
      return node.text
    end
    if node.t == "Emph" then
      return find_text_content(node.content[1])
    end
    if node.t == "Strong" then
      return find_text_content(node.content[1])
    end
    if node.t == "Strikeout" then
      return find_text_content(node.content[1])
    end
    if node.t == "Superscript" then
      return find_text_content(node.content[1])
    end
    if node.t == "Subscript" then
      return find_text_content(node.content[1])
    end
    if node.t == "SmallCaps" then
      return find_text_content(node.content[1])
    end
    if node.t == "Quoted" then
      return find_text_content(node.content[1])
    end
    if node.t == "Cite" then
      return find_text_content(node.content[1])
    end
    if node.t == "Code" then
      return node.text
    end
    if node.t == "Math" then
      return node.text
    end
    if node.t == "RawInline" then
      return node.text
    end
    if node.t == "Link" then
      return find_text_content(node.content[1])
    end
    if node.t == "Image" then
      return find_text_content(node.content[1])
    end
    if node.t == "Note" then
      return find_text_content(node.content[1])
    end
    if node.t == "Span" then
      return find_text_content(node.content[1])
    end
    if node.t == "RawBlock" then
      return node.text
    end
    if node.t == "Div" then
      return find_text_content(node.content[1])
    end
    if node.t == "BulletList" then
      return find_text_content(node.content[1])
    end
    if node.t == "OrderedList" then
      return find_text_content(node.content[1])
    end
    if node.t == "DefinitionList" then
      return find_text_content(node.content[1])
    end
    if node.t == "Table" then
      -- FIXME
      return nil
    end
    if node.t == "SimpleTable" then
      -- FIXME
      return nil
    end
    error("Unknown node type: " .. node.t)
    return nil
  end

  if tbl.caption ~= nil then
    local m = nil
    local which = nil
    if tbl.caption.long ~= nil then
      which = find_text_content(tbl.caption.long)
      if which == nil then
        return nil
      end
    elseif tbl.caption.short ~= nil then
      -- this won't match { .classes } or {.class1 .class2}
      -- it only matches {.class}, {.class1.class2}, etc. NO spaces!
      which = find_text_content(tbl.caption.long)
      if which == nil then
        return nil
      end
    end
    if which == nil then
      return nil
    end
    m = string.match(which[1].text, "{%.(.+)}")
    if m ~= nil then
      for _, class in ipairs(mysplit(m, ".")) do
        table.insert(tbl.classes, class)
      end
      table.move(which, 2, #which, 1)
      return tbl
    end
  end
end
```

**Reply by AlFontal** [reply]
Sorry, @cscheid, but am I supposed to be able to add classes to tables using @jmgirard's syntax if I add this as a lua filter (and load it in the document)?

In my case, it causes an error:
```
Error running filter _extensions/tables_filter.lua:
_extensions/tables_filter.lua:28: attempt to index a nil value (local 'node')
stack traceback:
        _extensions/tables_filter.lua:137: in function 'Table'
```

**Reply by cscheid** [reply]
We need a full reproducible example to be able to track down the crash.

**Reply by AlFontal** [reply]
@cscheid I mean, just creating a new default quarto project, saving your code in a file as `test_filter.lua` and then loading it into the `about.qmd` document with:
```yaml
title: "About"
filters:
  - test_filter.lua
```

And trying to render the documents prompts that same error. Changes in the actual document don't seem to matter at all. I'm completely oblivious to how to parse `lua` though, so I'm doing this kind of blindly at this point.

**Reply by cscheid** [reply]
Oh, I think I have a typo. I wrote that on the go, didn't test it:

```

    elseif tbl.caption.short ~= nil then
      -- this won't match { .classes } or {.class1 .class2}
      -- it only matches {.class}, {.class1.class2}, etc. NO spaces!
      which = find_text_content(tbl.caption.long) -- THIS SHOULD BE ".short"
      if which == nil then
        return nil
      end
```

**Reply by AlFontal** [reply]
Hi @cscheid!

Thanks a lot for this.

I fixed that typo, then I was still getting the error:

```
Error running test_filter.lua:
test_filter.lua:28: attempt to index a nil value (local 'node')
```

So, kind of blindly, assumed that the error came from exactly what the error says, indexing the `node` variable without previously checking if it was `nil`. So basically, I added this clause at the very beginning of the `find_text_context` function definition:

```lua
function Table(tbl)
  local function find_text_content(node)
    if node == nil then
        return nil
    end
   ...

```

This now works, and classes written in the caption are injected into the `<table>` element. However, since it's of course a caption, it is printed, which becomes rather annoying when it's just a string full of subclasses. Since this is just a hacky solution I defined an extra class `.hide-caption` so that in `styles.css` I define: 

```css
table.hide-caption  caption {
    display: none;
}
```

This seems to work to hide the caption. Of course it's problematic if you actually want to show the caption, but I guess a more elaborate lua filter might do the trick!

In any case, I know I might now be pushing it too far when this was just supposed to be a hacky trick to get over the limitation at this point, but a new issue arises:
Whenever a link is included in the table, the filter stops working, and no class is injected at all.

See:

```markdown

<!-- No link -->
| A | B | C |
|---|---|---|
| 1 | 2 | 3 |
| 4 | 5 | 6 |
| 7 | 8 | 9 |
| a | b | c |
| d | e | f |
: {.table-striped.table-hover.hide-caption}

<!-- Link -->

| A | B | C |
|---|---|---|
| 1 | 2 | 3 |
| 4 | 5 | 6 |
| 7 | 8 | 9 |
| a | b | c |
| [d](https://quarto.org) | e | f |
: {.table-striped.table-hover.hide-caption}

```

<img width="597" alt="imagen" src="https://user-images.githubusercontent.com/10958332/204238325-ccdfd260-c7b9-459f-8dfc-017311920558.png">

Any ideas on what might be causing the filter to break when a link is present in the table's text?

Thanks! And I completely understand if this is a non-priority.

**Reply by AlFontal** [reply]
Small update: html style links _do_ work. It's only markdown style `[text](url)` that seems to break it.

**Reply by cscheid** [reply]
@AlFontal Yes. I should have clarified that my solution was truly a _hack_. We have a better fix coming up in the near-to-medium term (cc @rich-iannone). At present, if you need full markdown rendering, you'll need to learn your way around Lua filters. In this case, you'll need https://pandoc.org/lua-filters.html#pandoc.read

**Reply by AlFontal** [reply]
Yes, understood! The hacky solution works well enough for my use-case now, and I ended up learning quite a bit about lua-filters! really looking forward the next releases, so many features coming up!

Thanks for the great work guys!
